scylla-cdc-printer/src/printer.rs

```diff
@@ -0,0 +1,273 @@
+        let fetcher = GenerationFetcher::new(&self.session);
+        let mut generation = self.get_first_generation(&fetcher).await?;
+
+        self.readers = vec![];
+        for stream_id in fetcher
+            .fetch_stream_ids(&generation)
+            .await?
+            .iter()
+            .flatten()
+        {
+            let reader = StreamReader::new(
+                &self.session,
+                vec![stream_id.clone()],
+                self.start_timestamp,
+                self.window_size,
+                self.safety_interval,
+                self.sleep_interval,
+            );
+            self.readers.push(Arc::new(reader));
+        }
+
+        fn print_row(row: Row) {
+            println!("{}", row.columns.len());
+        }
+
+        let mut futures: FuturesUnordered<_> = self
+            .readers
+            .iter()
+            .map(|reader| {
+                let reader = Arc::clone(reader);
+                let keyspace = keyspace.clone();
+                let table_name = table_name.clone();
+                tokio::spawn(async move { reader.fetch_cdc(keyspace, table_name, print_row).await })
+            })
+            .collect();
+
+        let (generation_sender, mut generation_receiver) = mpsc::channel(1);
+
+        let sleep_interval = self.sleep_interval;
+        let end_timestamp = self.end_timestamp;
+        tokio::spawn(async move {
+            generation = loop {
+                match fetcher.fetch_next_generation(&generation).await {
+                    Ok(Some(generation)) => break generation,
+                    Ok(None) => {
+                        let now = SystemTime::now()
+                            .duration_since(SystemTime::UNIX_EPOCH)
+                            .unwrap();
+                        if now + sleep_interval >= end_timestamp.to_std().unwrap() {
+                            break GenerationTimestamp {
+                                timestamp: end_timestamp,
+                            };
+                        }
+                        sleep(sleep_interval).await
+                    }
+                    Err(_) => {
+                        // TODO Log this properly
+                        println!("WARNING: Error during fetch_next_generation")
+                    }
+                }
+            };
+            generation_sender.send(generation).await.unwrap();
+        });
+
+        loop {
+            tokio::select! {
+                evt = futures.select_next_some() => {
+                    match evt {
+                        Ok(Err(_)) | Err(_) => {
+                            self.set_upper_timestamp(chrono::Duration::min_value()).await
+                        }
+                        _ => {}
+                    }
+                }
+                generation = generation_receiver.recv() => {
+                    self.set_upper_timestamp(generation.unwrap().timestamp).await;
+                }
+
+            }
+
```

Comment:
Here, you should wait until all `futures` are finished. We must wait until the streams are consumed to the end before we start reading next generation.

Next, if there was an error returned from any of the tasks, you should return it, too. Otherwise, you should create new stream readers and go back to the start of the loop.

Reply:
Does it matter if we wait for the futures to finish right here where you have mentioned it or just after we realise there exists a new generation, or is it simply a stylistic choice?

Reply:
What I meant is that you need to wait for all `futures` to finish after you `set_upper_timestamp` for all stream readers. AFAIK we have two cases here: either we realized there is a new generation or we observed that one of the `futures` returned an error.

I suggested placing the waiting code here because it looked like it would cover both of those cases. However, if you are rearranging the logic then it may make sense to place it somewhere else.